File: ts/src/calc.test.ts
import { describe, it, expect } from 'vitest'
import { add, sub, mul, divide, classify, clamp, isValidPort, withUnit, grade, isSentinel } from './calc'

describe('calc', () => {
  // Deliberately use only the below-range case: `n < lo` short-circuits, so
  // the `n > hi` branch remains unvisited on this executed source line.
  it('clamps below range', () => {
    expect(clamp(-3, 0, 10)).toBe(0)
  })

  // Each test below exercises only a subset of the branch outcomes on the
  // corresponding one-liner, keeping those lines PARTIALLY covered.
  it('rejects a non-integer port', () => expect(isValidPort(1.5)).toBe(false))
  it('formats with an explicit unit', () => expect(withUnit(3, 's')).toBe('3 s'))
  it('grades an A', () => expect(grade(95)).toBe('A'))
  // New in this PR: only n === 0 is passed, so the `|| n === 999` operand is
  // NEVER reached and the `if (n === 0 || n === 999)` line stays PARTIAL.
  it('detects the zero sentinel', () => expect(isSentinel(0)).toBe(true))

  it('adds', () => expect(add(2, 3)).toBe(5))
  it('subtracts', () => expect(sub(5, 3)).toBe(2))
  it('multiplies', () => expect(mul(4, 3)).toBe(12))
  it('divides', () => expect(divide(10, 2)).toBe(5))
  it('throws on divide by zero', () => expect(() => divide(1, 0)).toThrow('divide by zero'))

  it('classifies', () => {
    expect(classify(0)).toBe('zero')
    expect(classify(5)).toBe('positive')
    // classify(negative) intentionally untested → keeps coverage < 100%.
  })
})

File: go/calc_test.go
package calc

import "testing"

func TestAdd(t *testing.T) {
	if Add(2, 3) != 5 {
		t.Fatalf("Add(2,3) = %d, want 5", Add(2, 3))
	}
}

func TestDivide(t *testing.T) {
	q, err := Divide(10, 2)
	if err != nil || q != 5 {
		t.Fatalf("Divide(10,2) = %d, %v, want 5, nil", q, err)
	}
}

func TestDivideByZero(t *testing.T) {
	if _, err := Divide(1, 0); err == nil {
		t.Fatal("Divide(1,0) = nil error, want ErrDivideByZero")
	}
}

func TestClassify(t *testing.T) {
	if got := Classify(0); got != "zero" {
		t.Fatalf("Classify(0) = %q, want zero", got)
	}
	if got := Classify(5); got != "positive" {
		t.Fatalf("Classify(5) = %q, want positive", got)
	}
	// Newly covered in this PR: the negative branch was uncovered on main, so it
	// renders as "gained coverage" in the comparison.
	if got := Classify(-5); got != "negative" {
		t.Fatalf("Classify(-5) = %q, want negative", got)
	}
	// Covers the new "one" case added in this PR — "new covered".
	if got := Classify(1); got != "one" {
		t.Fatalf("Classify(1) = %q, want one", got)
	}
}

func TestPow(t *testing.T) {
	if Pow(2, 10) != 1024 {
		t.Fatalf("Pow(2,10) = %d, want 1024", Pow(2, 10))
	}
}

func TestIsSentinel(t *testing.T) {
	// Only n == 0 is passed, so `n == 999` is never reached in `if n == 0 || n == 999`.
	if !IsSentinel(0) {
		t.Fatal("IsSentinel(0) = false, want true")
	}
}

func TestWeakHash(t *testing.T) {
	if WeakHash("x") == "" {
		t.Fatal("WeakHash returned empty string")
	}
}
